### ISSUE TITLE
installation of package in temporary directory doesn't work as expected

### ISSUE THREAD

**Issue by brucekaufman** [pnpm/pnpm#3151]
I'm not sure if this is a bug or a feature request because I don't have a good understanding of what is fully supported by pnpm vs. npm currently.  I am investigating whether it is possible to migrate our serverless monorepo from npm to pnpm, because there are a number of compelling benefits to using pnpm in a monorepo with services that share many common dependencies.  To bundle up and deploy our packages, we rely on [serverless webpack](https://github.com/serverless-heaven/serverless-webpack), which is layer ontop of webpack for serverless applications.  By making a few small changes to the source code of serverless webpack, I've been able to enable pnpm and the most basic deployment cases are working!

However, I have run into an issue with packaging "externals".  These are dependency modules which have some incompatibility and so can't be bundled into the single js file that webpack creates.  Serverless webpack allows specifying externals which it than manages using an algorithm as follows:
1.  Create a temporary directory with a new package.json file which ONLY includes the externals as dependencies
2.  run npm install to create the `node_modules` folder for these dependencies
3.  the node_modules folder is included in the final .zip package and deployed along with the bunded js file produced by webpack

I tried to implement by changing step 3 to use `pnpm install .` to run the installation on the new temporary directory that i created.  However, the final structure in `node_modules` doesn't seem to match what i would expect.  Here is an example of a package.json file created by step 2 of the above:

```
{
  "name": "marley-sms-js",
  "version": "1.0.0",
  "description": "Packaged externals for marley-sms-js",
  "private": true,
  "scripts": {},
  "dependencies": {
    "@dazn/lambda-powertools-http-client": "1.28.1",
    "node-emoji": "^1.10.0",
    "twilio": "^3.28.0"
  }
}
```

and here is the final directory structure of node_modules in the new temporary directory created by pnpm:
```
./node_modules/marley-sms-js
./node_modules/@dazn/lambda-powertools-http-client
./node_modules/node-emoji
./node_modules/twilio
```
from my limited understanding, there are a two issues here:
1.  pnpm is not installing the dependencies of these modules, only the modules themselves
2. for some reason pnpm creates a new circular reference to the project itself (marley-sms-js contains the exact same package lock as the top level).   On 2nd installation I get the following error because of this circular reference:

```
 ERROR  ENAMETOOLONG: name too long, scandir '/Users/bruce.kaufman/Development/backend-pnpm/services/entities/interfaces/sms/marley-sms-js/.webpack/dependencies/node_modules/marley-sms-js/node_modules/marley-sms-js/node_modules/marley-sms-js/node_modules/marley-sms-js/node_modules/marley-sms-js/node_modules/marley-sms-js/node_modules/marley-sms-js/node_modules/marley-sms-js/node_modules/marley-sms-js/node_modules/marley-sms-js/node_modules/marley-sms-js/node_modules/marley-sms-js/node_modules/marley-sms-js/node_modules/marley-sms-js/node_modules/marley-sms-js/node_modules/marley-sms-js/node_modules/marley-sms-js/node_modules/marley-sms-js/node_modules/marley-sms-js/node_modules/marley-sms-js/node_modules/marley-sms-js/node_modules/marley-sms-js/node_modules/marley-sms-js/node_modules/marley-sms-js/node_modules/marley-sms-js/node_modules/marley-sms-js/node_modules/marley-sms-js/node_modules/marley-sms-js/node_modules/marley-sms-js/node_modules/marley-sms-js/node_modules/marley-sms-js/node_modules/twilio/lib/rest/preview/hosted_numbers/authorizationDocument'
```
Related to this issue, now in the package.json pnpm has added the dependency as follows:

```
    "marley-sms-js": "link:",
```

Basically I think my requirement is that I need to be able to create a new temporary directory with a package.json in a workspace and have the node_modules created as expected.  For some reason that doesn't happen here and I don't know if this is a bug or a new feature.  Please let me know how i can provide additional clarification to help understand the problem.  I can attempt to create a new repository with a minimal reproducible case but that will take some work so I was hoping that wouldn't be neccessary.

**Comment by zkochan** [comment]
What hosting do you use for your serverless app? As far as I know AWS doesn't work with symlinks and pnpm currently heavily relies on symlinks. In that case, you won't be able to use the node_modules created by pnpm at the moment.

> pnpm is not installing the dependencies of these modules, only the modules themselves

the dependencies of those modules are in the virtual store, inside `node_modules/.pnpm`.

I am not sure why `  "marley-sms-js": "link:",` was added to your package.json. Do you use a `pnpmfile.js`?

**Comment by brucekaufman** [comment]
Yes, we are using AWS lambda.  So perhaps my use case isn't viable currently.  It is really a shame because the installation time of the monorepo drops substantially and in general with the bundler I am able to create deployment packages that work in AWS. The issue is how handle these exception cases which need to be packaged traditionally, with hard copies of source plus all dependencies within the `node_modules` folder.  It would nice if there was some way to create this style of structure in some way for deployments so that I could solve this usage case. I could actually elect to use npm to handle creating these "externals" packages for deployment but use pnpm for the local file structure of the repo.  This feels strange but would actually work from what I can tell.  In general I am very impressed with pnpm and would love to use it but unless I can solve this problem in a robust way it doesn't look to be in the cards.

I do think there may be some bug in pnpm in this usage case but given the fact that I can't actually use it for the purpose I am intending I don't know if you want me to provide more detail or not.   One additional odd thing i noticed is that the node_modules folder for this temporary package is missing the `.pnpm` folder which includes all the links to dependencies as you mention.  All of the normal packages in the monorepo have this folder from what I can tell.  We are not using a pnpmfile.js either.

**Comment by zkochan** [comment]
Inside a workspace, only the root node_modules will contain the `node_modules/.pnpm` directory. All other node_modules will only contain symlinks to this central virtual store directory.

**Comment by mjyocca** [comment]
> Yes, we are using AWS lambda. So perhaps my use case isn't viable currently. It is really a shame because the installation time of the monorepo drops substantially and in general with the bundler I am able to create deployment packages that work in AWS. The issue is how handle these exception cases which need to be packaged traditionally, with hard copies of source plus all dependencies within the `node_modules` folder. It would nice if there was some way to create this style of structure in some way for deployments so that I could solve this usage case. I could actually elect to use npm to handle creating these "externals" packages for deployment but use pnpm for the local file structure of the repo. This feels strange but would actually work from what I can tell. In general I am very impressed with pnpm and would love to use it but unless I can solve this problem in a robust way it doesn't look to be in the cards.
> 
> I do think there may be some bug in pnpm in this usage case but given the fact that I can't actually use it for the purpose I am intending I don't know if you want me to provide more detail or not. One additional odd thing i noticed is that the node_modules folder for this temporary package is missing the `.pnpm` folder which includes all the links to dependencies as you mention. All of the normal packages in the monorepo have this folder from what I can tell. We are not using a pnpmfile.js either.
> 
> To summarize what serverless webpack is doing, it boils down to this sequence:
> 
> 1. Create a new subfolder inside a monorepo project `.webpack/dependencies` with a package.json as above
> 2. run `pnpm install .` in said folder

@brucekaufman were you able to get it all working with `serverless-webpack` or did you figure something else out? Facing similar issues currently.
